File: leetcode2/0142. Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        node=ListNode(0)
        node.next=head
        pslow,pfast=node,node
        while pfast and pfast.next and pfast.next.next:
            pslow=pslow.next
            pfast=pfast.next.next
            if pslow==pfast:
                pslow=node
                while pslow!=pfast:
                    pslow=pslow.next
                    pfast=pfast.next
                return pslow
        return None

File: C#/0142. Linked List Cycle II.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode DetectCycle(ListNode head) {
        if(head==null || head.next==null){
            return null;
        }
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode pslow=node.next;
        ListNode pfast=node.next.next;
        while(pfast.next!=null && pfast.next.next!=null && pslow!=pfast){
            pslow=pslow.next;
            pfast=pfast.next.next;
        }
        if(pfast.next==null || pfast.next.next==null){
            return null;
        }
        
        
        ListNode p=node;
        while(p!=pslow){
            p=p.next;
            pslow=pslow.next;
        }
        return pslow;
    }
}
